app/StateToken.php

```diff
@@ -24,6 +27,73 @@ class StateToken extends Model
     protected $hidden = [
     ];
 
+    /**
+     * StateToken constructor. Always end up with a UUID
+     * @param array $attributes
+     */
+    public function __construct(array $attributes = [])
+    {
+        // If empty or used, replace with a uuid
+        parent::__construct($attributes);
+
+        // if it exists and is used
+        if (array_key_exists('uuid', $attributes)) {
+            if (self::isUsedToken($attributes['uuid'])) {
+                $this->uuid = self::generateUnusedToken();
+            }
+        } else {
+            // If it doesn't exist, make one too.
+            $this->uuid = self::generateUnusedToken();
+        }
+    }
+
+    /**
+     * Makes and checks for an unused token
+     *
+     * @return string
+    */
+    public static function generateUnusedToken()
+    {
+        do {
```

Comment:
do .. while loops make me nervous. Is there another way?

Reply:
Nice use of a do while :D, it's so rare that they're needed. I suppose this could be expressed using recursion too. Not sure how PHP deals with recursion